### PR TITLE
Support keyboard selection in SelectabledRegion

### LINKED ISSUE #104541: [Selection] SelectionArea keyboard arrow key shortcut does not work
## Steps to Reproduce
```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: SelectionArea(
        child: Scaffold(
          appBar: AppBar(title: const Text(_title)),
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: const <Widget>[
                Text('Row 1'),
                Text('Row 2'),
                Text('Row 3'),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

1.  use mouse to select some text
3. try to use arrow key to change the selection

no of them work at the moment

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/104541

added two new selection events, and implement the handling method in renderparagraph and SelectionContainerDelegate.

GranularlyExtendSelectionEvent: handles shift+modifier+left/right
DirectionallyExtendSelectionEvent: handles shift + up/down

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by LongCatIsLooong — COMMENTED** [review]
I think I need to understand how seletable region works before I can do a full review.

File: packages/flutter/lib/src/rendering/selection.dart [hunk 1]
```diff
@@ -375,6 +399,106 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+  const DirectionallyExtendSelectionEvent({
+    required this.horizontalBaseline,
+    required this.isExtent,
+    required this.movement,
+  }) : super._(SelectionEventType.directionallyExtendSelection);
+
+  /// The horizontal offset the selection should move to.
+  ///
+  /// The offset is in global coordinates.
+  final double horizontalBaseline;
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:479]: this is the x axis coordinate when moving the selection line by line. I can't think of a better name.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 2]
```diff
@@ -830,6 +840,64 @@ class _SelectableRegionState extends State<SelectableRegion> with TextSelectionD
     await Clipboard.setData(ClipboardData(text: data.plainText));
   }
 
+  bool? _adjustingSelectionExtent;
+  bool _determineIsAdjustingSelectionExtend(bool forward) {
+    if (_adjustingSelectionExtent != null) {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:845]: The cache is here so that we will be moving the same selection handle once decided until user starts a new selection(a clear selection will be called).

**Comment by chunhtai** [comment]
Thanks @LongCatIsLooong I addressed all the comments, PTAL

**Comment by chunhtai** [comment]
a friendly bump

File: packages/flutter/lib/src/rendering/selection.dart [hunk 3]
```diff
@@ -375,6 +399,106 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+  previousLine,
+
+  /// Extends the selection to a certain horizontal offset in the next line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.horizontalBaseline] on how to
+  /// calculate the horizontal offset.
+  nextLine,
+
+  /// Extends the selection to by moving the selection edges forward to a
+  /// certain horizontal offset in the same line.
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:444]: I can't think of a case it would be in a different line. I can add more description.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 4]
```diff
@@ -830,6 +840,62 @@ class _SelectableRegionState extends State<SelectableRegion> with TextSelectionD
...
+    final bool isReversed;
+    final SelectionPoint start = _selectionDelegate.value
+        .startSelectionPoint!;
+    final SelectionPoint end = _selectionDelegate.value.endSelectionPoint!;
+    if (start.localPosition.dy > end.localPosition.dy) {
+      isReversed = true;
+    } else if (start.localPosition.dy < end.localPosition.dy) {
+      isReversed = false;
+    } else {
+      isReversed = start.localPosition.dx > end.localPosition.dx;
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:857]: I just tested with chrome, regardless of the text direction, forward=true will always move the selection handle that are lower in screen order. forward =false move the one one higher in screen order. so the logic here should sill hold.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 5]
```diff
@@ -1466,6 +1516,155 @@ class _SelectableFragment with Selectable, ChangeNotifier {
...
+        _textSelectionEnd ??= movement == DirectionalMovement.forward
+          ? TextPosition(offset: range.start)
+          : TextPosition(offset: range.end, affinity: TextAffinity.upstream);
+        _textSelectionStart ??= _textSelectionEnd;
+        final TextPosition targetedEdge = isExtent ? _textSelectionEnd! : _textSelectionStart!;
+        final Offset edgeOffsetInParagraphCoordinates = paragraph._getOffsetForPosition(targetedEdge);
+        final Offset baselineOffsetInParagraphCoordinates = Offset(
+          baselineInParagraphCoordinates,
+          // Use half of line height to point to the middle of the line.
+          edgeOffsetInParagraphCoordinates.dy - paragraph._textPainter.preferredLineHeight / 2
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/paragraph.dart:1559]: Missing a trailing comma here.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:845]: I am not sure if I understand. why end always represent the active edge given that start can be larger or smaller?

File: packages/flutter/lib/src/rendering/selection.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:479]: renamed to dx

File: packages/flutter/lib/src/rendering/selection.dart [hunk 6]
```diff
@@ -375,6 +399,106 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+    required this.forward,
+    required this.isExtent,
+    required this.granularity,
+  }) : super._(SelectionEventType.granularlyExtendSelection);
+
+  /// Whether to extend the selection forward.
+  final bool forward;
+
+  /// Whether this event is updating the end selection edge.
+  final bool isExtent;
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:420]: done

File: packages/flutter/lib/src/rendering/selection.dart [hunk 7]
```diff
@@ -375,6 +399,106 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+
+  /// The granularity for which the selection extend.
+  final TextGranularity granularity;
+}
+
+/// The directional to extend a selection.
+///
+/// The [DirectionallyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
+enum DirectionalMovement {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:430]: renamed to SelectionExtendDirection

File: packages/flutter/lib/src/rendering/selection.dart [hunk 8]
```diff
@@ -296,6 +296,30 @@ enum SelectionEventType {
...
+
+  /// An event that extends the selection.
+  directionallyExtendSelection,
+}
+
+/// The unit of how selection handles move in text.
+///
+/// The [GranularlyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
+enum TextGranularity {
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/selection.dart:311]: CC @Renzo-Olivares since this reminds me of your work on longpress dragging by word and triple click dragging by line.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 9]
```diff
@@ -1596,4 +1795,25 @@ class _SelectableFragment with Selectable, ChangeNotifier {
       );
     }
   }
+
+  @override
+  TextSelection getLineAtOffset(TextPosition position) {
+    final TextRange line = paragraph._getLineAtOffset(position);
+    final int start = line.start.clamp(range.start, range.end); // ignore_clamp_double_lint
+    final int end = line.end.clamp(range.start, range.end); // ignore_clamp_double_lint
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/paragraph.dart:1803]: Why can't you use clampDouble for these?

File: packages/flutter/lib/src/rendering/selection.dart [hunk 10]
```diff
@@ -296,6 +296,30 @@ enum SelectionEventType {
...
+  granularlyExtendSelection,
+
+  /// An event that extends the selection in a specific direction.
+  directionallyExtendSelection,
+}
+
+/// The unit of how selection handles move in text.
+///
+/// The [GranularlyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/selection.dart:310]: Nit: "their" => "its"

File: packages/flutter/lib/src/rendering/selection.dart [hunk 11]
```diff
@@ -375,6 +399,114 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+
+/// The directional to extend a selection.
+///
+/// The [DirectionallyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
+enum SelectionExtendDirection {
+  /// Move one edge of the selection vertically to the previous adjacent line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/selection.dart:434]: Nit: If you think it would be helpful, maybe clarify that this means the previous word-wrapped line, and not the previous hard linebreak?  Same for nextLine.

File: packages/flutter/lib/src/rendering/selection.dart [hunk 12]
```diff
@@ -375,6 +399,114 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+
+  /// Move the selection edges forward to a certain horizontal offset in the
+  /// same line.
+  ///
+  /// If there is no on-going selection, the selection must start with the first
+  /// line (or equivalence of first line in non-text selectable) and select
+  /// toward the horizontal offset in the same line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/selection.dart:451]: What if the horizontal offset is extends beyond the width of the field, does it wrap to the next line or stop at this line?

Maybe here isn't the place to explain that, but just something I was wondering while reading these docs.

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 13]
```diff
@@ -1250,6 +1250,116 @@ class _ScrollableSelectionContainerDelegate extends MultiSelectableSelectionCont
     return result;
   }
 
+  @override
+  SelectionResult handleGranularlyExtendSelection(GranularlyExtendSelectionEvent event) {
+    final SelectionResult result = super.handleGranularlyExtendSelection(event);
+    // The selection geometry may not have the accurate offset for the edges
+    // that are outside of the viewport whose transform may not be valid. Only
+    // the edge this event is updating is sure to be accurate.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/scrollable.dart:1258]: That's tricky, I'm glad you documented it.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 14]
```diff
@@ -242,6 +242,15 @@ class _SelectableRegionState extends State<SelectableRegion> with TextSelectionD
   late final Map<Type, Action<Intent>> _actions = <Type, Action<Intent>>{
     SelectAllTextIntent: _makeOverridable(_SelectAllAction(this)),
     CopySelectionTextIntent: _makeOverridable(_CopySelectionAction(this)),
+    ExtendSelectionToNextWordBoundaryOrCaretLocationIntent: _makeOverridable(_GranularlyExtendSelectionAction<ExtendSelectionToNextWordBoundaryOrCaretLocationIntent>(this, granularity: TextGranularity.word)),
+    ExpandSelectionToDocumentBoundaryIntent: _makeOverridable(_GranularlyExtendSelectionAction<ExpandSelectionToDocumentBoundaryIntent>(this, granularity: TextGranularity.document)),
+    ExpandSelectionToLineBreakIntent: _makeOverridable(_GranularlyExtendSelectionAction<ExpandSelectionToLineBreakIntent>(this, granularity: TextGranularity.line)),
+
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/selectable_region.dart:248]: Why the newline here?

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 15]
```diff
@@ -830,6 +840,61 @@ class _SelectableRegionState extends State<SelectableRegion> with TextSelectionD
     await Clipboard.setData(ClipboardData(text: data.plainText));
   }
 
+  bool? _adjustingSelectionExtent;
+  bool _determineIsAdjustingSelectionExtend(bool forward) {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/selectable_region.dart:844]: "Extend" => "Extent"?

File: packages/flutter/test/rendering/paragraph_test.dart [hunk 16]
```diff
@@ -821,12 +822,12 @@ void main() {
...
         selectable.dispatchSelectionEvent(
           SelectionEdgeUpdateEvent.forStart(
-            globalPosition: paragraph.getOffsetForCaret(start, Rect.zero),
+            globalPosition: paragraph.getOffsetForCaret(start, Rect.zero) + const Offset(0, 5),
           ),
         );
         selectable.dispatchSelectionEvent(
           SelectionEdgeUpdateEvent.forEnd(
-            globalPosition: paragraph.getOffsetForCaret(end, Rect.zero),
+            globalPosition: paragraph.getOffsetForCaret(end, Rect.zero) + const Offset(0, 5),
```
Comment (justinmc) [inline packages/flutter/test/rendering/paragraph_test.dart:830]: Why did you need to add that small y offset?

File: packages/flutter/test/rendering/paragraph_test.dart [hunk 17]
```diff
@@ -911,6 +912,392 @@ void main() {
...
+      // Equivalent to sending shift + arrow-right
+      registrar.selectables[0].dispatchSelectionEvent(
+        const GranularlyExtendSelectionEvent(
+          forward: true,
+          isEnd: true,
+          granularity: TextGranularity.character,
+        ),
+      );
+      selection = paragraph.selections[0];
+      expect(selection.start, 4); // how `ar`e you
```
Comment (justinmc) [inline packages/flutter/test/rendering/paragraph_test.dart:947]: I think elsewhere in our tests I have seen `how [ar]e you` or `how |ar|e you`.  Maybe we should standardize at some point.

No action needed though, just thinking out loud.

File: packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart [hunk 18]
```diff
@@ -290,8 +290,8 @@ class DefaultTextEditingShortcuts extends StatelessWidget {
 
     const SingleActivator(LogicalKeyboardKey.arrowLeft, shift: true, alt: true): const ExtendSelectionToNextWordBoundaryOrCaretLocationIntent(forward: false),
     const SingleActivator(LogicalKeyboardKey.arrowRight, shift: true, alt: true): const ExtendSelectionToNextWordBoundaryOrCaretLocationIntent(forward: true),
-    const SingleActivator(LogicalKeyboardKey.arrowUp, shift: true, alt: true): const ExtendSelectionToLineBreakIntent(forward: false, collapseSelection: false, collapseAtReversal: true),
-    const SingleActivator(LogicalKeyboardKey.arrowDown, shift: true, alt: true): const ExtendSelectionToLineBreakIntent(forward: true, collapseSelection: false, collapseAtReversal: true),
+    const SingleActivator(LogicalKeyboardKey.arrowUp, shift: true, alt: true): const ExtendSelectionVerticallyToAdjacentLineIntent(forward: false, collapseSelection: false),
+    const SingleActivator(LogicalKeyboardKey.arrowDown, shift: true, alt: true): const ExtendSelectionVerticallyToAdjacentLineIntent(forward: true, collapseSelection: false),
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart:294]: I guess this was incorrect?

File: packages/flutter/test/widgets/scrollable_selection_test.dart [hunk 19]
```diff
@@ -11,6 +11,27 @@ import 'package:flutter_test/flutter_test.dart';
 
 import 'clipboard_utils.dart';
 
+Future<void> sendKeyCombination(
```
Comment (justinmc) [inline packages/flutter/test/widgets/scrollable_selection_test.dart:14]: Maybe we should get rid of [`sendKeys` in editable_text_test](https://github.com/flutter/flutter/blob/b20b7d995b17c130dcd7456556cd69c1fde5e858/packages/flutter/test/widgets/editable_text_test.dart#L5581) and just use this.  I like the idea of just using a SingleActivator.

Comment (justinmc) [reply packages/flutter/test/widgets/scrollable_selection_test.dart:14]: I see that you added a copy of this function to another test file.  It looks like this is already public though, maybe just add it to a shared _utils.dart file?

**Review by justinmc — COMMENTED** [review]
Reading this really makes me wish that we could unify the selection code here and in RenderEditable.  I think this logic is partially duplicated in EditableText's Action handlers.  Maybe the core problem is the RenderEditable/RenderParagraph divide.  It is nice that this can receive the same Intents from DefaultTextEditingShortcuts at least, though.

It's also hard to be confident that our text selection keyboard shortcuts work identically in EditableText and SelectableRegion.  I wonder if it would be possible to write a test that could be reused by both?  Like sending keyboard keys should be the same for both, and both could produce a TextSelection where the same `expect`s could be used.  Then you could write a test that goes through all of typical keyboard shortcuts for modifying selection (not editing) and use it for both.

I'm just daydreaming though, it might not be feasible to write something like that in this PR...

**Comment by chunhtai** [comment]
> Reading this really makes me wish that we could unify the selection code here and in RenderEditable

That is my plan once the selectable region is at least feature on part with the EditableText.

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 9 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart:1803]: This is in type, so we should not use clampDouble.
 
the purpose of this ignore_clamp_double_lint is to make sure we DON'T use the .clamp method on `double`. The problem is that our current lint can't distinguish between .clamp called on `int` or `double` or any `num` subclass. So the current suggestion is to add this ignore if you are sure the clamp is called on non-double data type.

File: packages/flutter/lib/src/rendering/selection.dart (hunk 12 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/selection.dart:451]: I don't think we can define how to handle these cases here. It depends on the selectable they may choose to select things differently

File: packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart (hunk 18 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart:294]: no they were not

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/selectable_region.dart]: According to https://api.flutter.dev/flutter/services/TextSelection/extent.html 
> When the user uses the arrow keys to adjust the selection, this is the value that changes. 

Does it make sense to do the same here?

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart]: Is the document up to date? I thought even for TextField this may not be the case, it will pick the one that will expand the selection?

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/selectable_region.dart:924]: Ah ok, I think macOS does the same for bidi when moving the caret (but that may not be true for deletion).  Maybe rename this to left/right instead of forward/backward? I think in editable text land I used forward/backward for logical direction. When we unify the API it will be nice to make that consistent.

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:924]: I renamed extent to end. I am using start and end to describe the handles. Do you think this is good enough? If you meant the input parameter `forward` I think that is accurate since it represent arrow right/arrow down.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 15 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart]: I renamed it to End which is more accurate

File: packages/flutter/test/rendering/paragraph_test.dart (hunk 16 above)
Comment (chunhtai) [reply packages/flutter/test/rendering/paragraph_test.dart:830]: originally it was selecting right at the edge which can select across line if there is multi line text. It was ok because there was not multi line tests in this file prior

File: packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart (hunk 18 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart:299]: Ok same behavior, different Intent :+1:

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 20]
```diff
@@ -1027,6 +1033,32 @@ class RenderParagraph extends RenderBox
     return _textPainter.getWordBoundary(position);
   }
 
+  TextRange _getLineAtOffset(TextPosition position) => _textPainter.getLineBoundary(position);
+
+  TextPosition _getTextPositionAbove(TextPosition position) {
+    // The caret offset gives a location in the upper left hand corner of
+    // the caret so the middle of the line above is a half line above that
+    // point and the line below is 1.5 lines below that point.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/paragraph.dart:1041]: Super nit here and below:  You mention the caret here, but there's nothing in this method directly related to the caret (I see that you're using getOffsetForCaret in _getTextPositionVertical, but it's not obvious looking only at this method why you mention the caret).

Also similarly, it's not clear in this method why you're trying to find a point in the middle of the line (I guess you just need a point that's going to give you what you want from getPositionForOffset).

File: packages/flutter/lib/src/rendering/selection.dart [hunk 21]
```diff
@@ -375,6 +399,118 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
...
+  final bool forward;
+
+  /// Whether this event is updating the end selection edge.
+  final bool isEnd;
+
+  /// The granularity for which the selection extend.
+  final TextGranularity granularity;
+}
+
+/// The directional to extend a selection.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/selection.dart:426]: "directional" => "direction"

File: packages/flutter/test/widgets/keyboard_utils.dart [hunk 22]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+Future<void> sendKeyCombination(
```
Comment (justinmc) [inline packages/flutter/test/widgets/keyboard_utils.dart:9]: Looks good, thank you for taking the time to make this reusable!

**Review by justinmc — APPROVED** [review]
LGTM :+1: 

Thanks for the updates.

**Review by LongCatIsLooong — COMMENTED** [review]
LGTM for the most part.

Question: what order are `selectables` in? I guess what I don't understand is that why we can hand the event to the next selectable in the list instead of having to look for the next closest selectable in that direction?

**Comment by chunhtai** [comment]
The selectables are in screen order by default.
